pandas/core/indexes/multi.py

```diff
@@ -3690,10 +3690,9 @@ def _union(self, other, sort) -> MultiIndex:
             )
 
         else:
-            rvals = other._values.astype(object, copy=False)
-            right_missing = lib.fast_unique_multiple(self._values, rvals)
-            if right_missing:
-                result = self.append(other.take(right_missing))
+            right_missing = other.difference(self)
```

Comment:
How does this behave with duplicates and especially with Indexes that contain tuples?

Apart from that, I would prefer not doing that now. I have a different pr that I would prefer merging first.

Reply:
Duplicates shouldn’t reach this given the if else block this is within (other is always unique). Not sure about tuples but I can take a look. Or I can just close it if you have something in the works.

Reply:
Sure, I'll take a look at the duplicates. I'll leave this open a bit longer.

Reply:
@phofl, are you open to this one now that the duplicates case has been optimized a bit?

Reply:
Can you rebase now?